pkg/controller/common/tracing/apmclientgo/client.go

```diff
@@ -0,0 +1,182 @@
+	}
+
+	propagateLegacyHeader := tx.ShouldPropagateLegacyHeader()
+	requestName := requestName(req)
+	name := spanName(requestName)
+	span := tx.StartSpan(name, "db.kubernetes", apm.SpanFromContext(ctx))
+
+	if span.Dropped() {
+		span.End()
+		apmhttp.SetHeaders(req, traceContext, propagateLegacyHeader)
+		return r.r.RoundTrip(req)
+	}
+
+	traceContext = span.TraceContext()
+	ctx = apm.ContextWithSpan(ctx, span)
+	req = apmhttp.RequestWithContext(ctx, req)
+	span.Context.SetHTTPRequest(req)
+	span.Context.SetDestinationService(apm.DestinationServiceSpanContext{
+		Name:     "Kubernetes API server",
+		Resource: "Kubernetes",
+	})
+	span.Context.SetDatabase(apm.DatabaseSpanContext{
+		Statement: requestName,
+		Type:      "kubernetes",
+	})
+
+	apmhttp.SetHeaders(req, traceContext, propagateLegacyHeader)
+	resp, err := r.r.RoundTrip(req)
+	if err != nil {
+		span.End()
+	} else {
+		span.Context.SetHTTPStatusCode(resp.StatusCode)
+		resp.Body = &responseBody{span: span, body: resp.Body}
+	}
+	return resp, err
+}
+
+// CloseIdleConnections calls r.r.CloseIdleConnections if the method exists.
+func (r *roundTripper) CloseIdleConnections() {
+	type closeIdler interface {
+		CloseIdleConnections()
+	}
+	if tr, ok := r.r.(closeIdler); ok {
+		tr.CloseIdleConnections()
+	}
+}
+
+// CancelRequest calls r.r.CancelRequest(req) if the method exists.
+func (r *roundTripper) CancelRequest(req *http.Request) {
+	type cancelRequester interface {
+		CancelRequest(*http.Request)
+	}
+	if r, ok := r.r.(cancelRequester); ok {
+		r.CancelRequest(req)
+	}
+}
+
+type responseBody struct {
+	span *apm.Span
+	body io.ReadCloser
+}
+
+// Close closes the response body, and ends the span if it hasn't already been ended.
+func (b *responseBody) Close() error {
+	b.endSpan()
+	return b.body.Close()
+}
+
+// Read reads from the response body, and ends the span when io.EOF is returned if
+// the span hasn't already been ended.
+func (b *responseBody) Read(p []byte) (n int, err error) {
+	n, err = b.body.Read(p)
+	if errors.Is(err, io.EOF) {
+		b.endSpan()
+	}
+	return n, err
+}
+
+func (b *responseBody) endSpan() {
+	addr := (*unsafe.Pointer)(unsafe.Pointer(&b.span))
+	if old := atomic.SwapPointer(addr, nil); old != nil {
+		(*apm.Span)(old).End()
+	}
+}
+
+func spanName(reqName string) string {
+	const prefix = "Kubernetes:"
+	var b strings.Builder
+	b.Grow(len(prefix) + 1 + len(reqName))
+	b.WriteString(prefix)
+	b.WriteRune(' ')
+	b.WriteString(reqName)
+	return b.String()
+}
+
+func requestName(req *http.Request) string {
+	statement := req.Method
+	numSegments := 2
+	// add a bit more context in the summary for PUT requests e.g. namespace
+	if req.Method == "PUT" {
+		numSegments = 3
+	}
+
+	pathSegments := strings.Split(req.URL.Path, "/")
+	path := strings.Join(pathSegments[len(pathSegments)-numSegments:], "/")
+	// let's call out watch requests explicitly
+	if watch := req.URL.Query().Get("watch"); watch == "true" {
+		statement = "WATCH"
+	}
```

Comment:
Any reason to not just use `req.URL.Path`?

<img width="1941" alt="image" src="https://user-images.githubusercontent.com/976373/169487667-a4d1b738-1964-4e1c-a162-b7bfc3fd870c.png">

Reply:
I thought it got very long in the UI so I was looking for something shorter that still conveyed the most relevant information.